b_cfn_elasticsearch_cloner/initial_cloner_source/action.py

```diff
@@ -0,0 +1,142 @@
+import json
+import logging
+import os
+from decimal import Decimal
+from typing import Dict, Any, Optional, Tuple
+
+import boto3
+from elasticsearch import Elasticsearch, RequestsHttpConnection
+from elasticsearch.serializer import JSONSerializer
+from elasticsearch.helpers import bulk
+from requests_aws4auth import AWS4Auth
+
+
+logger = logging.getLogger(__name__)
+
+dynamodb = boto3.resource("dynamodb")
+
+class Action:
+    def __init__(self, invocation_event: Dict[str, Any]):
+        self.__invocation_event: Dict[str, Any] = invocation_event
+        self.__parameters: Dict[str, Any] = invocation_event["ResourceProperties"]
+        self.__old_parameters: Dict[str, Any] = invocation_event["OldResourceProperties"]
+        self.__resource_id: Optional[str] = invocation_event.get("PhysicalResourceId")
+
+        try:
+            self.REGION = os.environ["AWS_REGION"]
+
+            self.TABLE_NAME = self.__parameters["DynamodbTableName"]
+            self.ELASTICSEARCH_ENDPOINT = self.__parameters["ElasticsearchEndpoint"]
+            self.INDEX_NAME = self.__parameters["ElasticsearchIndexName"]
+        except KeyError as ex:
+            logger.error(f"Missing environment: {repr(ex)}.")
+            raise
+
+    def create(self) -> Tuple[Optional[Dict[Any, Any]], Optional[str]]:
+        """
+        Clones existing DynamoDB table data to an Elasticsearch index.
+
+        :return: A tuple containing two items:
+            1. Custom data to return back to CloudFormation service.
+            2. Physical resource id (can be empty).
+        """
+        logger.info(f"Initiating resource creation with these parameters: {json.dumps(self.__parameters)}.")
+
+        credentials = boto3.Session().get_credentials()
+        awsauth = AWS4Auth(
+            credentials.access_key,
+            credentials.secret_key,
+            self.REGION,
+            "es",
+            session_token=credentials.token,
+        )
+
+        class DynamodbEncoder(JSONSerializer):
+            def default(self, data):
+                if isinstance(data, set):
+                    return list(data)
+                if isinstance(data, Decimal):
+                    return float(data)
+                return JSONSerializer.default(self, data)
+
+        es = Elasticsearch(
+            hosts=[{"host": self.ELASTICSEARCH_ENDPOINT, "port": 443}],
+            http_auth=awsauth,
+            use_ssl=True,
+            verify_certs=True,
+            connection_class=RequestsHttpConnection,
+            serializer=DynamodbEncoder(),
+        )
+
+        dynamodb_table = dynamodb.Table(self.TABLE_NAME)
+
+        id_field = self.__get_dynamodb_primary_key(dynamodb_table)
+
+        bulk(
```

Comment:
This function also handles main logic. Please add just a few comments for those who do not know what ES bulk function does. A URL reference should be great too.